### ISSUE TITLE
protocol for new mixins

### ISSUE THREAD

**Issue by janrito** [typeddjango/djangorestframework-stubs#302]
It would be nice if we could use the classes already defined in `generics.py` as protocols.

For example, if I want to define a new viewset mixin

```python
class TagModelMixin:
    @action(
        detail=True,
        methods=["PUT"],
        url_path="tag/(?P<slug>[^/.]+)",
    )
    def tag(self, request: Request, slug: str, pk: Any = None) -> Response:
        """Tag an instance with a specific tag"""
        instance = self.get_object()  
        instance.tag.add(slug)
        ...
```

mypy complains that `TagModelMixin` does not have a `get_object` attribute

I think the solution is to define a protocol for the api of `GenericAPIView` - which is sort of already defined in `generics.py`

**Comment by Viicos** [comment]
What if you specify `self` as being a type of a `GenericAPIView`? e.g. `def tag(self: GenericAPIView, request: Request, slug: str, pk: Any = None) -> Response:`

iirc this is the recommended way of doing by `mypy`, although it is with protocols, so not sure it will work as excepted with a real class.

**Comment by Viicos** [comment]
Yes I believe protocols is the only possibility to correctly type mixins. As it is pretty common to use mixins in drf (and django as well), this could be implemented as a helper protocol

**Comment by PedroPerpetua** [comment]
I personally use this snippet:

```py
"""Type alias for Mixins."""
if TYPE_CHECKING:
    GenericViewMixin = GenericAPIView
    APIViewMixin = APIView
else:
    GenericViewMixin = object
    APIViewMixin = object
```

And have my Mixins inherit from the corresponding class.
